torchprime/torch_xla_models/model/model_utils.py

```diff
@@ -78,3 +198,222 @@ def extract_model_size_from_model_name(model_name: str) -> int | float:
+    elif "embed" in name:
+      param_groups["embedding"] += param.numel()
+    elif "lm_head" in name:
+      param_groups["lm_head"] += param.numel()
+    elif "norm" in name or "layernorm" in name:
+      param_groups["norm"] += param.numel()
+    else:
+      param_groups["other"] += param.numel()
+
+  for k, v in param_groups.items():
+    percentage = (v / total_params) * 100
+    logger.info("  {:10s}: {} params ({:.2f}%)".format(k, f"{v:,}", percentage))
+
+
+def get_param_group_key(param_name: str) -> str:
+  """Return a deterministic *shard key* for any parameter name.
+
+  Heuristics (in priority order):
+
+  1. Strip a leading ``module.`` (used by `nn.DataParallel` / `FSDP` wraps).
+  2. Transformer-style: ``{model,encoder,decoder}.layers.<idx>.…`` ➜ ``layers_<idx>``.
+  3. Generic block pattern ``<block>.<idx>.…`` (e.g. ``layer1.0.*``,
+     ``down_blocks.3.*``) ➜ ``<block>_<idx>``.
+  4. Fallback to the first path component (``lm_head``, ``bias`` …).
+
+  The returned string is short (good for filenames) and stable across runs.
+  """
+  parts = param_name.split(".")
+
+  # (1) Strip wrapper prefixes inserted by some wrappers
+  if parts[0] == "module":
+    parts = parts[1:]
+
+  # (2) Transformer blocks (model.layers.N / encoder.layers.N / decoder.layers.N)
+  if (
+    len(parts) >= 3
+    and parts[0] in {"model", "encoder", "decoder"}
+    and parts[1] == "layers"
+    and parts[2].isdigit()
+  ):
+    return f"layers_{parts[2]}"
+
+  # (3) Generic "<block>.<idx>.*" – covers ResNet, UNet, Swin, ConvNeXt …
+  if len(parts) >= 2 and parts[1].isdigit():
+    return f"{parts[0]}_{parts[1]}"
+
+  # (4) Last-chance fallback
+  return parts[0]
+
+
+def save_distributed_checkpoint(model: torch.nn.Module, save_dir: Path) -> None:
+  """Save the model state using torch.distributed.checkpoint (DCP).
+
+  Args:
+    model: The model whose state_dict should be saved.
+    save_dir: Directory where the distributed checkpoint will be written.
+  """
+  state_dict = {"model": model.state_dict()}
+  dist_cp.save(
+    state_dict=state_dict,
+    storage_writer=dist_cp.FileSystemWriter(
+      str(save_dir), thread_count=max(2, min(8, mp.cpu_count()))
+    ),
+    planner=xc.SPMDSavePlanner(),
+  )
+  logger.info("DCP checkpoint written to %s", save_dir)
+
+
+def convert_to_safetensors_on_cpu(model: torch.nn.Module, save_dir: Path) -> None:
+  """Reload checkpoint on CPU and export sharded safetensors files.
+
+  Assumes the checkpoint has already been written. This method should only be run on Rank 0.
+  Exaple usage:
+  ```python
+  if xr.process_index() == 0:
+    model_utils.convert_to_safetensors_on_cpu(...)
+  ```
+
+  Args:
+    model: The model used to construct the CPU placeholder state_dict.
+    save_dir: Directory where the original checkpoint is saved.
+              Safetensors files and index will also be written here.
+  """
+  logger.info("Rank-0: reloading checkpoint for safetensors export …")
```

Comment:
nit: you only enforce rank0 here through comment on the method, which is a weak assumption.
I'd move this to the place where method is called from, or just remove rank0 from log